### DISCUSSION TITLE
Combine whole diagram into flat article

### DISCUSSION THREAD

**Discussion by macin** [rpggio/obsidian-chat-stream#20] [Ideas]
The way this plugin allows me to chat with chatgpt is a game changer. However, once I'm done, I'd love to be able with a single click to transfer visual notes into an article with headers and content. This would make much more sense when sharing with others as visual format can easily get to large to share as pdf and be understandable by the receipient.

**Comment by rpggio** [comment]
@macin Thanks for the feedback. A feature could be added to compile selected canvas notes into a single file note. Would this work? What were you thinking of RE headers?

**Reply by VentureTactics** [reply]
The suggestion there to combine selected nodes would be great.
It's sort of like turning the nodes into a 'thread' type document to store the conversation.

**Reply by rpggio** [reply]
@VentureTactics the nodes could be copied into a document note (leaving canvas unchanged), or they could be converted to ref nodes, where the document becomes the source and each note is a snippet from the document note. Which better matches the use cases you're thinking of?

**Reply by macin** [reply]
I'd add from my perspective... In an ideal solution, putting technical limitations aside, I'd love to keep both in sync: document snippet and canvas note. That means if I have a doc and I find something to fix in a snippet, then I can see that correction in the canvas note and run ChatGPT on the full path including what I changed in the particular note... The same for the other way rounds if I decide to rewrite the note on canvas, I'd like it to be reflected in the document snippet. Would it be possible @rpggio ?

**Reply by VentureTactics** [reply]
@rpggio If we went with the ref nodes option, my concern would be in how that affects using existing but unrelated nodes for context -- particularly when branches are created in a thread and components are re-used.

For example, Say I had a chat thread like this:
<Thread 1>
SystemPrompt1 -> Command1 -> AssistantOutput1 ->Command 2 -> AssistantOutput2 -> [MERGE]
</Thread 1>

Then a second thread in the same chat using some of the previous nodes:
<Thread 2>
SystemPrompt1 -> Command1 -> AssistantOutput1 -[branch]-> Command3 -> AssistantOutput3
</Thread 2>

My understanding is that it will use context of an entire file that's linked in the thread.

So if the whole thread became just links in a single document, and components were re-used, wouldn't it contain unwanted content from Command2 and AssistantOutput2? If so, it would mess up the context, unless it was only pulling in the referenced headings being linked.

I do see value in Macin's suggestion of keeping things in sync in their original format, but if there has to be a single source of truth... I'd say the nodes should remain as they are with content copied into a document with the ability to re-merge changes (unless one of the nodes is already a file, in which case just the link reference will probably suffice).

One of the reasons this is important is that Canvas files are far less supported than standard MD files.

For instance, maybe a particular task would be better suited for another plugin like 'Text Generator' or Copilot, in which case Chat Stream does some initial work, merges the selected canvas nodes together into a MD file, then Copilot can be run on that file in a seperate step.

I wonder how %Templater% can be used with this in some way to assign outputs to the right spot in a template document... but that may be a problem for another day. ;)

**Reply by rpggio** [reply]
I agree about the better support for MD notes. I have already received requests about adding features from those other AI plugins. Making it easier to migrate or sync that content between canvas and MD seemed a more useful route.

Chat Stream will only include the referenced content from a ref node. So, you can cite content from an MD doc in a canvas node, and only that content will be used. If it's acting otherwise, that would be a regression bug.

It seems you both are in favor of MD notes being the content source, if the content is to be duplicated or shared with canvas.

Migrating a set of canvas notes into an MD doc, or generating a canvas from multiple selections within an MD doc are both interesting features, which are not really AI-specific at all.

**Reply by VentureTactics** [reply]
Merging the nodes probably seems un-related to AI in the most direct sense.

It seems like this would actually be a very important part of chaining prompts together and actually utilizing the output.

Here's a workflow to illustrate why this feature is pretty important:
![Chat Stream Merge Relevance Illustration](https://github.com/rpggio/obsidian-chat-stream/assets/88736327/31de753b-692d-474d-abad-531dec9ea638)

**Reply by rpggio** [reply]
@VentureTactics 
I'm thinking that a new model could provide the flexibility to build these more complex flows. What if we introduced command nodes? A node following particular /syntax would be considered a command node. 
Ideas:

/complete - Runs the current Chat Stream behavior, with ancestors as context. (Could accept an overwrite param to specify whether target note is overwritten)
/merge - Merge the ancestor nodes into a single node. If the target node is a file node, the file will be overwritten on each run.
/expand - Expand {{template}} slots from named arrow inputs
/explode - Converts input note/file into separate output nodes, breaking on headings
/run - Run all command nodes that this note points to (via dotted arrow?)

I'm interested to know how this concept intersects with other plugins, and how well it might suit your use cases.

**Reply by VentureTactics** [reply]
@rpggio a command node would be a GREAT way to accomplish this.

Every one of those suggestions makes perfect sense. I'll attempt to modify my fictional workflow with those in mind and will share it here when done.

With this kind of command node, it may also be possible to include a /webhook command with Post and GET  capability (and standard auth options like Bearer), with the response emitting to the next /Await node.

This way information can be retrieved into or sent from Obsidian in an actual workflow -- which would unlock endless ways of modifying the AI context, even with information not yet stored in Obsidian.

For example...
You could create a Chat Stream workflow that queries your Quickbooks via API and returns this months transaction data, then the AI is asked to create a summary (likely with other node datapoints as well), and the /expand step merges the summary into a selected frontmatter property in the following file (a file that's reference on a seperate financial dashboard canvas in Obsidian).

### Re: Plugins
Commander Plugin (or similar) could be used to manually run workflows or schedule chron runs as long as there was a clear way to trigger the /run (maybe we can give the run a workflow name to identify it?)

Metadata Menu is what I use to manage my frontmatter properties. 

Dataview is my preferred way to query on and display frontmatter. If Chat Stream can run a workflow and replace the frontmatter of a file as directed, dv tables can be used to always display the up-to-date property (also aided by chron runs where applicable).

Templater could possibly be used to merge data into templates, however, if we use the arrow connectors to push into a specific frontmatter property, something like Templater can be used from there (referencing the properties alone) without ever actually interacting with the Chat Stream plugin.

Advanced Canvas is also a must for me, and I do utilize the Minimal Theme off and on (which is why I have a different color scheme for my canvas nodes).

**Reply by VentureTactics** [reply]
Alright @rpggio, take a look at this concept workflow and see if it's in line with what you're thinking with these advanced workflow commands: 👍
  
![Chat Stream Advanced Command Node Concept Workflow](https://github.com/rpggio/obsidian-chat-stream/assets/88736327/ecbfa3dd-2656-471d-ae39-c9c675757062)

You'll notice it also contains example elements from some of the other idea threads like adding a model weight node and assigning assistant nodes.

**Reply by rpggio** [reply]
I'm thinking that commands be limited to command nodes. Command can show a run/stop button, show 'running' status, etc. Do you think that constraint is workable?

Output label param is a good idea - that can be specified as frontmatter prop.

I like the colored arrows showing which are system-generated.

I'll have to think more about assistant-targeted use cases. 
Have you tried creating a [GPT assistant](https://platform.openai.com/docs/assistants/overview?context=with-streaming)? (different than custom GPT) They have API-managed threads and file uploads. Not sure yet how those OpenAI assistants intersect with this tool.

If you haven't picked up TS or Python, you should start. That is where your experiment is heading. This is evident from what you're diagramming.

**Reply by VentureTactics** [reply]
Hah, I haven't picked up TS or Python yet. In fact, I can't code at all.
I can loosely understand what's going on in python when I see it, though I rely on GPTs for coding tasks at present.
Having said that, I'm extremely proficient with LLM prompting and no-code/low-code workflows -- as you might imagine. ;)
I also have a knack for data and complex business strategy.

**Reply by rpggio** [reply]
@VentureTactics, I think the way to implement canvas command nodes would be with a new plugin.
Canvas Commander would enable you to create command nodes. The available commands could be extended by users or by other plugins.
Chat Stream could register commands with Canvas Commander automatically, or CCommander could have its own ancestor walking and AI completion commands.
Commands like your http-request command could be implemented as extensions.

It might be worth adding executable canvas nodes as feature request for the core team, to see if it gains any traction. I'll take a look.

**Reply by rpggio** [reply]
Added [feature request](https://forum.obsidian.md/t/canvas-command-nodes/80138).

**Reply by VentureTactics** [reply]
A separate plugin would probably be perfect.
Good idea.
That way it can serve more bespoke use cases.

**Reply by VentureTactics** [reply]
Came across a few repos today which could be relevant in some of the aforementioned workflow steps.

### Merging and un-merging canvas nodes:
For merging nodes into a single md file:
https://github.com/slnsys/obsidian-canvas2document

An alternative approach to applying canvas nodes to a document (or series of documents):
https://github.com/aarongilly/obsidian-semantic-canvas-plugin

For splitting the note back up:
https://github.com/trey-wallis/obsidian-note-splitter

### Replacing pre-set variables in a md file:
For embedding a file's content into a persistent variable:
https://github.com/nates-foo/obsidian-dynamic-templates

An alternative implementation of variables (but not persistent) relying instead on one-time replacement of an inline embed with the actual content of the embedded file (instead of just a reference):
https://github.com/maxiee/obsidian-knowledge-weaver

@rpggio maybe take a look at these repos and see if there might be synergy with a few of them, or code that can help accomplish some of the workflow goals.

**Comment by macin** [comment]
Oh no, I missed your questions @rpggio , sorry for that!
What I meant was just take the notes as they are, assuming the user will take care of putting headers in each of the note. I'd keep it simple...
